nixos/lib/make-disk-image.nix: support content mode and ownership

fgaz opened the pull request:
###### Motivation for this change

Annoyance 1 in #23052

###### Things done

* [x] Tested by building a custom image with additional contents

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

fgaz committed:
nixos/lib/make-disk-image.nix: support content mode and ownership

fgaz commented:
/marvin opt-in
/status awaiting_reviewer

fgaz commented:
@timokau I think I broke it by assigning manually the first reviewer...

I don't see a way to stop it, so I'll just:

/status needs_reviewer

Ekleog commented on nixos/lib/make-disk-image.nix:
```diff
@@ -174,22 +178,31 @@ let format' = format; in let
+      # Unfortunately cptofs only supports modes, not ownership, so can't use
+      # rsync's --chown option. Instead, we change the ownerships in the
+      # VM script with chown.
+      rsync_flags="-a --no-o --no-g $rsync_chmod_flags"
```
Hmm… I think there's a problem here, because you don't reset `rsync_chmod_flags` to `""` between two loop runs.

fgaz replied:
right... mutability strikes again >.<

will fix.

Ekleog commented on nixos/lib/make-disk-image.nix:
```diff
@@ -284,6 +297,21 @@ in pkgs.vmTools.runInLinuxVM (
+        group="''${groups_[$i]}"
+        if [ -n "$user$group" ]; then
+          # We have to nixos-enter since we need to use the user and group of the VM
+          nixos-enter --root $mountPoint -- chown -R "$user:$group" "$target"
```
Looks to me like this will behave unexpectedly if in particular only `user` is specified. Maybe add an assert to prevent cases where only one of user/group is being set?

fgaz replied:
`chown` from `coreutils` supports both `user:` and `:group`. Do we have to support other `chown`s that don't allow that syntax?

Ekleog replied:
`:group` appears to work correctly, but when I tried `chown root: test-file` it also changed the group of `test-file` to `root`.

TBH, I'm not sure we have to support other `chown`s that don't allow that syntax, but it feels to me like restricting our API surface to say “you're allowed to chown iff you set both user and group” wouldn't be a big problem in practice (as most of the time one would want to do both), and it might help avoiding unintuitive behavior?

fgaz replied:
You're right, I missed this part of the chown docs:

> If a colon but no group name follows the user name, that user is  made  the
> owner  of  the files **and the group of the files is changed to that user's login group**

At this point couldn't we just add another if and mirror the etc.*.uid/gid behaviour? I don't think this will have any problematic cases:

```sh
if [ -n $group ]; then
  chown $user:$group $file
else
  if [ -n $user ]; then
    chown $user $file
  fi
fi
```

Ekleog replied:
That's totally possible too, if you think there's an actual use case for people setting only one of user/group — I'm not sure it'd bring much value and it expands the API we then have to maintain, but why not.

fgaz replied:
I have no real use case in mind — I'll leave the API as is and document the limited options then.

Ekleog reviewed (changes requested):
Looks like I can't directly comment on a non-modified line, but this probably deserves documentation, at least around the `contents ? []` line, and maybe at other places if we already have documentation about make-disk-image.

Also, do you know of an easy way to test this PR, ideally something directly nix-buildable like a NixOS test?

Ekleog commented:
Oh and also, wrt.
> I think I broke it by assigning manually the first reviewer...

AFAIU the bot currently doesn't ever actually assign a new reviewer (because of a current bug in the implementation), so you probably didn't break it, it's just that it's a reason why marvin is still beta :)

timokau commented:
Yeah, it's a bug: https://github.com/timokau/marvin-mk2/issues/99

I looked into the issue a couple of days ago and it seemed to be a bit more work than I thought. Still not a lot, but my schedule is completely packed recently. So if the bot is spamming you, just reset it manually for now. Sorry about that.

Ekleog commented:
/status awaiting_changes

(anyone understands why marvin changed the status back to awaiting_reviewer? I can't see any new commit nor marvin command since the review, so this looks weird to me)

fgaz commented:
/status awaiting_changes

I think further comments (either from you or me) after a rewiew reset the status to awaiting_reviewer

timokau commented:
> I think further comments (either from you or me) after a rewiew reset the status to awaiting_reviewer

Only comments from the PR author ([implementation](https://github.com/timokau/marvin-mk2/blob/34c1b7fa98dd48044c4819f84f5fc98617a8c49b/marvin/status.py#L81)). Probably [this](https://github.com/NixOS/nixpkgs/pull/104292#discussion_r535455679) comment. Its maybe a bit of an over-eager heuristic. The intent is to err on the side of `awaiting_reviewer`, since nagging only happens in that status. `awaiting_changes` is kind of the "abandoned PR sink", so I want to avoid putting PRs there on accident. The status doesn't have to be entirely correct, it only really matters if there are three days without activity. Still, there is room for improvement.

timokau commented:
Although it looks like it kind of worked as intended here, since it re-ignited the previously stale discussion :)

Ekleog commented:
Makes sense!

(FWIW, I've been sick the last few days and didn't open the computer. I only noticed the nagging after having read the answers… that said even though it wasn't required it did work as intended here I guess, as it had been more than 3 days waiting for my feedback ^^')

timokau commented:
Yeah, 3 days might be a bit on the shorter side in general.

I hope you feel better already and will feel better still soon :)

fgaz commented:
@Ekleog 

* [x] documented
* [x] added the assertion
* [x] fixed the bug in the loop

I also removed the possibility to use the uid and gid attributes, to keep it simple.

fgaz commented:
> Also, do you know of an easy way to test this PR, ideally something directly nix-buildable like a NixOS test?

Hmm, no, sorry, my use case is external: https://git.sr.ht/~sircmpwn/builds.sr.ht/tree/master/images/nixos

Ekleog commented:
Thank you! This looks good to me otherwise, the only missing thing is ideally a test to ensure non-regression, though we could probably land as is anyway provided the current VM tests (eg. the openstack-image-* tests) pass, as they already test some previous parts of make-disk-image.

Feel free to let me know if you want to give up on adding a test to ensure non-regression, though that will mean this feature will be less stable than it could be :)

/status awaiting_changes

fgaz committed:
nixosTests.image-contents: init

fgaz commented:
@Ekleog I added the test. It's based on the ec2 stuff because it makes it fairly easy to test an actual image instead of a configuration, not sure if that's ok.

fgaz commented:
@GrahamcOfBorg build nixosTests.image-contents

Ekleog reviewed (approved):
Basing the test on the ec2 test infrastructure is perfectly good! Going to restart ofborg as it looks like the aarch64 test failed due to a machine being rebooted or similar

@GrahamcOfBorg build nixosTests.image-contents

Ekleog commented:
@GrahamcOfBorg build nixosTests.image-contents

Hmm… maybe asking ofborg directly from a review doesn't work?

Ekleog commented:
@GrahamcOfBorg build nixosTests.ec2-config

Let's see whether other tests made with makeEc2Test build on aarch64

Ekleog commented:
Wellll, looks like nixosTests.ec2-config worked on aarch64 but failed on x86_64. Let's just say makeEc2Test is probably wobbly and land this :)

If you have spare time, feel free to have a look as to why these tests don't consistently work on all architectures, but for now I'm calling this good enough and don't want to push more work on you. Thank you for all your work on this PR! :)

fgaz commented:
And thanks to you for the review!

Would it be OK to open a backport PR for this?

Ekleog commented:
Given it's a new feature and not a bugfix, in my opinion it's something that cannot be backported, especially seeing as it's risky given the fact it touches a quite core part of our system. Sorry! :/

fgaz commented:
I understand, I'll just vendor the file for now :)
